src/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs

```diff
@@ -0,0 +1,355 @@
+        public static readonly KnownHeader TransferEncoding = new KnownHeader("Transfer-Encoding", HttpHeaderType.General, TransferCodingHeaderParser.MultipleValueParser);
+        public static readonly KnownHeader Upgrade = new KnownHeader("Upgrade", HttpHeaderType.General, GenericHeaderParser.MultipleValueProductParser);
+        public static readonly KnownHeader UpgradeInsecureRequests = new KnownHeader("Upgrade-Insecure-Requests");
+        public static readonly KnownHeader UserAgent = new KnownHeader("User-Agent", HttpHeaderType.Request, ProductInfoHeaderParser.MultipleValueParser);
+        public static readonly KnownHeader Vary = new KnownHeader("Vary", HttpHeaderType.Response, GenericHeaderParser.TokenListParser);
+        public static readonly KnownHeader Via = new KnownHeader("Via", HttpHeaderType.General, GenericHeaderParser.MultipleValueViaParser);
+        public static readonly KnownHeader WWWAuthenticate = new KnownHeader("WWW-Authenticate", HttpHeaderType.Response, GenericHeaderParser.MultipleValueAuthenticationParser);
+        public static readonly KnownHeader Warning = new KnownHeader("Warning", HttpHeaderType.General, GenericHeaderParser.MultipleValueWarningParser);
+        public static readonly KnownHeader XAspNetVersion = new KnownHeader("X-AspNet-Version");
+        public static readonly KnownHeader XContentDuration = new KnownHeader("X-Content-Duration");
+        public static readonly KnownHeader XContentTypeOptions = new KnownHeader("X-Content-Type-Options");
+        public static readonly KnownHeader XFrameOptions = new KnownHeader("X-Frame-Options");
+        public static readonly KnownHeader XMSEdgeRef = new KnownHeader("X-MSEdge-Ref");
+        public static readonly KnownHeader XPoweredBy = new KnownHeader("X-Powered-By");
+        public static readonly KnownHeader XRequestID = new KnownHeader("X-Request-ID");
+        public static readonly KnownHeader XUACompatible = new KnownHeader("X-UA-Compatible");
+
+        // Helper interface for making GetCandidate generic over strings, utf8, etc
+        private interface IHeaderNameAccessor
+        {
+            int Length { get; }
+            char CharAt(int index);
+        }
+
+        private struct StringAccessor : IHeaderNameAccessor
+        {
+            private readonly string _string;
+
+            public StringAccessor(string s)
+            {
+                _string = s;
+            }
+
+            public int Length => _string.Length;
+            public char CharAt(int index) => _string[index];
+        }
+
+        // Find possible known header match via lookup on length and a distinguishing char for that length.
+        // Matching is case-insenstive.
+        // NOTE: Because of this, we do not preserve the case of the original header,
+        // whether from the wire or from the user explicitly setting a known header using a header name string.
```

Comment:
> We still need to treat non-standard casing as a known header, so that we know what parser to use and what type of header it is (content vs request etc).

I see.

> Are you concerned about perf here or about preserving the original case?

A little of both.  Not a big concern (and I believe today we already normalize the casing).